Consent container banner  tests

ChrisBAshton opened the pull request:
Resolves #1413 

**Overall change:** _Adding Cypress tests for the Cookie Banner on canonical pages ._

**Code changes:**

- _Testing the banner appears._
- _Testing that the banner disappears once relevant buttons have been clicked._
- _Testing that the cookies are assigned the relevant values._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

jamesbrumpton committed:
Adding Cookie Banner Cypress Tests

jamesbrumpton committed:
Adding Cookie Banner Cypress Tests

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner--tests

Commit:
Merge branch 'ConsentContainer_banner' of github.com:BBC-News/simorgh into ConsentContainer_banner--tests

dr3 commented on cypress.json:
```diff
@@ -1,6 +1,8 @@
+   "chromeWebSecurity": false,
   "env": {
     "APP_ENV": "local"
+
```
```suggestion
```

Commit:
Update cypress.json

dr3 commented on cypress/integration/cookieBanner.js:
```diff
@@ -0,0 +1,80 @@
+
+    retrieveCookieValue('ckns_policy', '000');
+  });
+});
```
I think we should add a test similar to
```
it('should not override the users default cookie policy', () => {
    cy.clearCookies();

    cy.setCookie('ckns_policy', '101');

    cy.visit(`/news/articles/${config.assets.newsThreeSubheadlines}`);

    retrieveCookieValue('ckns_policy', '101');

    cy.contains('No, take me to settings').click();

    retrieveCookieValue('ckns_policy', '101');
});
```

jamesdonoh replied:
Similarly, I think we are lacking a test for the scenario where a user has previously seen the privacy banner, but not the cookie banner.

jamesdonoh commented on cypress/integration/cookieBanner.js:
```diff
@@ -0,0 +1,80 @@
+    retrieveCookieValue('ckns_privacy', '1');
+    retrieveCookieValue('ckns_policy', '000');
+
+    cy.contains('OK').click();
```
Isn't there a chance this might end up clicking the wrong thing on the page, if something else contains the text "OK"? Was just wondering if we could use the element ID instead to be more precise.

jamesdonoh commented on cypress/integration/cookieBanner.js:
```diff
@@ -0,0 +1,80 @@
+    getPrivacyBanner().should('not.be.visible');
+
+    expectCookieExpiryDateOneYear('ckns_explicit');
+    expectCookieExpiryDateOneYear('ckns_policy');
```
Should we also check the expiry date of `ckns_privacy`? And should we verify the path and domain of the cookies to ensure consistency with Orbit?

ChrisBAshton replied:
Part 1: done.

Part 2: yes, would be a good idea. Trickier to implement as it changes per env. Perhaps better implemented in https://github.com/bbc/simorgh/pull/1427, where we'll be tackling per-env cookie oven tests.

jamesdonoh commented on cypress/integration/cookieBanner.js:
```diff
@@ -0,0 +1,80 @@
+    cy.contains('No, take me to settings').click();
+
+    expectCookieExpiryDateOneYear('ckns_explicit');
+    expectCookieExpiryDateOneYear('ckns_policy');
```
It would make it easier to read/compare these tests if the steps were executed in a consistent order - e.g. cookie expiry checks at the end, as in the previous one. Also perhaps an opportunity to factor out some of the duplication?

jamesdonoh commented on cypress/support/metaTestHelper.js:
```diff
@@ -6,6 +6,18 @@ export const testResponseCode = (path, responseCode) => {
   });
 };
 
+export const retrieveCookieValue = (cookieName, value) => {
```
This could be called `expectCookieValue` or something to make its functionality clearer.

jamesdonoh commented on cypress/support/metaTestHelper.js:
```diff
@@ -6,6 +6,18 @@ export const testResponseCode = (path, responseCode) => {
+  cy.getCookie(cookieName).should('have.property', 'value', value);
+};
+
+export const expectCookieExpiryDateOneYear = cookieName => {
```
I'd prefer `expectDefaultCookieExpiry` (or maybe, `expectCookieExpiry(cookieName, DEFAULT_EXPIRY)`) to avoid putting a business rule that might change into the function name.

jamesdonoh commented on cypress.json:
```diff
@@ -1,5 +1,6 @@
 {
   "video": false,
+   "chromeWebSecurity": false,
```
Indent slightly off.
```suggestion
  "chromeWebSecurity": false,
```
Also, I'd like to understand why we need to do this, as it seems like it could make us insensitive to some bugs we might want Cypress to help us know about. Is it because of the clicking on "Find out what's changed"?

ChrisBAshton replied:
Have just tried running locally without it, this is what happens:

```
     CypressError: Cypress detected a cross origin error happened on page load:

  > Blocked a frame with origin "http://localhost:7080" from accessing a cross-origin frame.

Before the page load, you were bound to the origin policy:
  > http://localhost:7080

A cross origin error happens when your application navigates to a new superdomain which does not match the origin policy above.

This typically happens in one of three ways:

1. You clicked an <a> that routed you outside of your application
2. You submitted a form and your server redirected you outside of your application
3. You used a javascript redirect to a page outside of your application

Cypress does not allow you to change superdomains within a single test.

You may need to restructure some of your test code to avoid this problem.

Alternatively you can also disable Chrome Web Security which will turn off this restriction by setting { chromeWebSecurity: false } in your 'cypress.json' file.

https://on.cypress.io/cross-origin-violation

      at Object.cypressErr (http://localhost:7080/__cypress/runner/cypress_runner.js:65283:11)
      at Object.throwErr (http://localhost:7080/__cypress/runner/cypress_runner.js:65248:18)
      at Object.throwErrByPath (http://localhost:7080/__cypress/runner/cypress_runner.js:65275:17)
      at onPageLoadErr (http://localhost:7080/__cypress/runner/cypress_runner.js:55386:21)
      at HTMLIFrameElement.<anonymous> (http://localhost:7080/__cypress/runner/cypress_runner.js:61113:19)
      at HTMLIFrameElement.dispatch (http://localhost:7080/__cypress/runner/cypress_runner.js:17523:27)
      at HTMLIFrameElement.elemData.handle (http://localhost:7080/__cypress/runner/cypress_runner.js:17335:28)
```

Doesn't look like there's a good alternative tbh.

jamesdonoh replied:
Do we know which navigation causes the error? If it's navigating by clicking on an `<a>` do we actually need to do that? Advice here https://docs.cypress.io/guides/guides/web-security.html#External-Navigation suggests asserting the existence of/target of the link rather than clicking.

ChrisBAshton replied:
Hmmm but we do two things with that - when we click on the link, we set a cookie, which we can then check the value of. If we only assert the `href` of the link, we can't assert that clicking it sets the cookie we expect.

jamesdonoh replied:
I see what you mean. It just seems a bit of a shame to permanently disable web security in the test suite just to test setting a cookie, which is not even the actual cause of the error. Is there an easy way we could stub out/remove the href instead? Or have the event handler return false? Just spitballing...

ChrisBAshton committed:
adds additional cookie edge-case test

ChrisBAshton committed:
fixes indentation of cypress.json file

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner--tests

ChrisBAshton committed:
_actually_ fixes indentation in cypress.json

Something went wrong in https://github.com/bbc/simorgh/pull/1439/commits/b6bd1507be4b752a665685d834d641a358e999ed

ChrisBAshton committed:
Merge branch 'ConsentContainer_banner' into ConsentContainer_banner--tests

ChrisBAshton committed:
renames retrieveCookieValue to assertCookieValue

Makes intent clearer

ChrisBAshton committed:
refactor cookie expiry functions according to code review

ChrisBAshton committed:
namespace cy.contains() calls and refactors one test

ChrisBAshton committed:
assert default cookie policy not overwritten on page load

ChrisBAshton committed:
refactor tests

dr3 reviewed (approved):
![](https://media.giphy.com/media/3oz8xRF0v9WMAUVLNK/source.gif)

Commit:
add the missing live fixture

Commit:
unit test values

ChrisBAshton committed:
adds test for returning users who haven't explicitly accepted cookies

dr3 commented on cypress/integration/cookieBanner.js:
```diff
@@ -0,0 +1,111 @@
+    ensureCookieExpiryDates();
+  });
+
+  it('should show cookie banner (and NOT privacy banner) if user has visited the page before and didn\'t explicitly "accept" cookies', () => {
```
```suggestion
  it("should show cookie banner (and NOT privacy banner) if user has visited the page before and didn't explicitly 'accept' cookies", () => {
```

ChrisBAshton replied:
Agreed 👍 have done a local commit, wasn't sure if linter would complain.

We should start using backticks everywhere IMO!

ChrisBAshton committed:
prefer double-quotes than escaping

Commit:
Merge pull request #1445 from bbc/add-missing-live-fixture

Add missing live fixture

dr3 commented on cypress/integration/cookieBanner.js:
```diff
@@ -0,0 +1,111 @@
+  });
+};
+
+describe('Article Body Tests', () => {
```
```suggestion
describe('Canonical Cookie Banner Tests', () => {
```

Commit:
rename test more descriptively

Co-Authored-By: ChrisBAshton <ChrisBAshton@users.noreply.github.com>

twinlensreflex reviewed (approved):
lgtm

ChrisBAshton committed:
refactor tests - make explicit when visiting article and setting cookies

ChrisBAshton committed:
Merge branch 'latest' into ConsentContainer_banner--tests

ChrisBAshton committed:
merge with ConsentContainer_banner--tests on remote

ChrisBAshton committed:
increase buffer to 15 milliseconds to allow for slow tests
